### PR TITLE
Config rework

### PR DESCRIPTION
**Pull request by enyst**
**Not for review yet.** It's partially backported (messy stuff) and Claude was playing too much. 😅 

This PR proposes a fix, the addition of a new small config for memory, and a few refactorings for readability:

- [x] fix regression when the user has custom llms defined, and an [llm] section: if some attributes are missing from a custom llm definition, but are defined in [llm], it should fallback to those in [llm].
- [x] break up the embeddings part from LlmConfig into a MemoryConfig
- [x] refactoring: we're starting to have too many classes and they have duplicate methods, introduce an ABC
- [x] refactoring: break up the long functions loading from toml and env, into loading per dataclass: each little config knows better what it needs to load and how to deal with some specifics.
- [ ] better documentation.

### CODE REVIEW

File: opendevin/core/config.py [hunk 1]
```diff
@@ -79,13 +115,53 @@ class LLMConfig:
...
+        llm_configs = {}
+        default_config = cls()
+
+        # Load default [llm] section if it exists
+        if 'llm' in toml_config and isinstance(toml_config['llm'], dict):
+            default_config = cls(**toml_config['llm'])
+            llm_configs['llm'] = default_config
+
+        # Load custom LLM configs, falling back to default for unspecified attributes
+        for key, value in toml_config.get('llm', {}).items():
```
Comment (tobitege) [inline opendevin/core/config.py:142]: This worries me that pulling values from one section into another introduces unexpected behaviors.
Sections should be standalone and not treated as inheritable classes, so to speak.
If a required value is missing, there just needs to be a clear error for the user to see.

File: opendevin/core/config.py [hunk 2]
```diff
@@ -211,38 +375,55 @@ class SandboxConfig(metaclass=Singleton):
...
 @dataclass
-class AppConfig(metaclass=Singleton):
+class AppConfig(BaseConfig, metaclass=SingletonABCMeta):
     """Configuration for the app.
 
     Attributes:
         llms: A dictionary of name -> LLM configuration. Default config is under 'llm' key.
+        memories: A dictionary of name -> Memory configuration. Default config is under 'memory' key.
         agents: A dictionary of name -> Agent configuration. Default config is under 'agent' key.
+        memories: A dictionary of name -> Memory configuration. Default config is under 'memory' key.
```
Comment (tobitege) [inline opendevin/core/config.py:426]: duplicate

**Comment by tobitege** [comment]
Great job tackling this! Having more docstrings alone will make a difference.

File: opendevin/core/config.py (hunk 2 above)
Comment (enyst) [reply opendevin/core/config.py]: Which one was a duplicate? `memories` is just one?

Comment (tobitege) [reply opendevin/core/config.py]: Yes, the "memories" line was it.

**Comment by enyst** [comment]
I thought I did `git pull upstream main`... 🤷 Maybe it got outdated again because y'all are working too hard! 😅

**Pull request closed without merging**